funsor/distributions.py

```diff
@@ -42,17 +48,11 @@ class DistributionMeta(FunsorMeta):
     """
     def __call__(cls, *args, **kwargs):
         kwargs.update(zip(cls._ast_fields, args))
-        args = cls._fill_defaults(**kwargs)
-        args = numbers_to_tensors(*args)
-
-        # If value was explicitly specified, evaluate under current interpretation.
-        if 'value' in kwargs:
-            return super(DistributionMeta, cls).__call__(*args)
-
-        # Otherwise lazily construct a distribution instance.
-        # This makes it cheaper to construct observations in minipyro.
-        with interpretation(lazy):
-            return super(DistributionMeta, cls).__call__(*args)
+        value = kwargs.pop('value', 'value')
+        kwargs = OrderedDict((k, to_funsor(v)) for k, v in kwargs.items())
```

Comment:
Did I miss a major change in Python 3.6 that makes kwargs deterministically ordered? Can you point me to any reading material about this change?

Reply:
Insertion order in dictionaries was [made deterministic in Python 3.6](https://docs.python.org/3/whatsnew/3.6.html#new-dict-implementation) and this was [added to the language spec in 3.7](https://docs.python.org/3/whatsnew/3.7.html) ("Python data model improvements").  I guess that's why this code isn't breaking.

Reply:
> The order-preserving aspect of this new implementation is considered an implementation detail and should not be relied upon

I'd feel safer if we avoided reliance on deterministic order for now (there being so many other potential pitfalls). How much extra work would it be to avoid relying on this?

Reply:
Shouldn't be much, I'll make all the suggested determinism-related changes

Reply:
Oh look at this, [PEP 468](https://www.python.org/dev/peps/pep-0468/) was [accepted](https://docs.python.org/3/whatsnew/3.6.html#pep-468-preserving-keyword-argument-order) in Python 3.6. So I guess we're safe :smile:

> **kwargs in a function signature is now guaranteed to be an insertion-order-preserving mapping.

Reply:
@eb8680 maybe I'm overreacting about the effort required to resume support for Python 3.5. This comment thread illustrates my main concern, that we would need to pay special attention to order in kwargs and other use of dictionaries.

Reply:
Python 3.5 compatibility in Funsor [has always been nominal at best](https://github.com/pyro-ppl/funsor/commit/986e75af9a3e32b65c2d2b5d4774ebcdbff5b9ab#diff-354f30a63fb0907d4ad57269548329e3R31).  I am happy to reopen and continue working on #329 but that only makes sense if we can agree beforehand on specific, objective criteria that would give us enough confidence to go ahead with adding Funsor as an optional dependency to Pyro.

Reply:
> nominal at best

Well early on we support Python 2.7 and we took care to ensure determinism. I think this PR on March 25 was the first where I stopped reviewing for dict order determinism.

> if we can agree beforehand on specific, objective criteria

Sounds reasonable. I think there are two concerns: ensuring determinism and ensuring correctness. I'm willing to give up determinism (at least across python processes; see [PYTHONHASHSEED](https://docs.python.org/3.3/using/cmdline.html#envvar-PYTHONHASHSEED) in Python 3.3+). But we'll need to avoid dangerous nondeterminism, e.g. 
```py
# This is correct only in Python 3.6+:
Tensor(torch.randn(2, 3), OrderedDict(a=bint(2), b=bint(3)))
```
Since this is difficult to test, I can only offer subjective criteria. To make this as objective as possible, I suggest we split your 3.5 PR #329 into two parts:
1. #329 objective: get existing tests to pass; followed by
2. #xxx subjective: we meet for 1 hour to decide subjective criteria in the form of say regular expressions we can run agains source code (e.g. `OrderedDict(\w+(\w\d)?=`). We code those up in a `test_source_code.py` and make a second PR that changes only enough code to satisfy those properties.

wdyt?

Reply:
Those are good suggestions, but I would prefer to avoid any further speculative coding.  To that end I suggest we reverse the order of operations and start a PR with a failing `test_source_code.py` and a list of any other show-stopping compatibility problems, ideally expressed as failing unit tests.  

If we can agree that fixing the problems in the PR is *both necessary and sufficient* to add Funsor as an optional dependency to Pyro, I can make the actual fixes including getting existing tests to pass in #329.  If we cannot agree, or the list appears to require too much engineering effort, then there's no reason to proceed with #329 or any related work.

If this plan is acceptable, I will put up a PR in the next few days with a seed version of `test_source_code.py`.

Reply:
That sounds like a reasonable plan. I'm happy to chat about `test_source_code.py` or just see what you come up with 🙂

It might help us discover dangerous patterns to simply run the tests under Python 3.5.